### PR TITLE
Update DataImportCron CronJob when needed

### PR DESCRIPTION
**Pull request by arnongilboa**
Signed-off-by: Arnon Gilboa <agilboa@redhat.com>

**What this PR does / why we need it**:
When CDI is being upgraded, the DataImportCron CronJobs are kept as is, and refer cdi-importer image of the previous version, rather than the current (upgraded) one.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
Update DataImportCron CronJob container image if needed due to upgrade
```

### CODE REVIEW

File: pkg/controller/dataimportcron-controller.go [hunk 1]
```diff
@@ -702,10 +711,19 @@ func addDataImportCronControllerWatches(mgr manager.Manager, c controller.Contro
 	return nil
 }
 
-func (r *DataImportCronReconciler) cronJobExists(ctx context.Context, cron *cdiv1.DataImportCron) bool {
+func (r *DataImportCronReconciler) cronJobExistsAndUpdated(ctx context.Context, cron *cdiv1.DataImportCron) (bool, error) {
 	var cronJob batchv1.CronJob
 	cronJobNamespacedName := types.NamespacedName{Namespace: r.cdiNamespace, Name: GetCronJobName(cron)}
-	return r.uncachedClient.Get(ctx, cronJobNamespacedName, &cronJob) == nil
```
Comment (akalenyu) [inline pkg/controller/dataimportcron-controller.go:708]: Why were we doing uncached client here btw?

File: pkg/controller/dataimportcron-controller.go [hunk 2]
```diff
@@ -702,10 +711,19 @@ func addDataImportCronControllerWatches(mgr manager.Manager, c controller.Contro
...
+	err := r.uncachedClient.Get(ctx, cronJobNamespacedName, &cronJob)
+	podSpec := &cronJob.Spec.JobTemplate.Spec.Template.Spec
+	if err == nil && len(podSpec.Containers) > 0 && podSpec.Containers[0].Image != r.image {
+		podSpec.Containers[0].Image = r.image
```
Comment (akalenyu) [inline pkg/controller/dataimportcron-controller.go:720]: Maybe Update() over newCronJob().Spec != foundCron.Spec?
It is possible at some point in the future we decided to change some fields in the cronjob object too, and want to sync existing cronjobs to those values

File: pkg/controller/dataimportcron-controller.go (hunk 1 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:708]: afair r.client.Get() sometime returned NotFound after creating CronJob. However, I'll try again r.client for Get() and Update() here and check if it's consistent.

File: pkg/controller/dataimportcron-controller.go (hunk 2 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:720]: @akalenyu tried this change and it seems that when CronJob is created based on newCronJob(), k8s patches several fields of the spec (dnsPolicy, schedulerName, serviceAccount etc.) so we cannot simply compare (DeepEqual) the current and the desired spec.

File: pkg/controller/dataimportcron-controller.go [hunk 3]
```diff
@@ -702,10 +711,19 @@ func addDataImportCronControllerWatches(mgr manager.Manager, c controller.Contro
...
+		err = r.uncachedClient.Update(ctx, &cronJob)
```
Comment (awels) [inline pkg/controller/dataimportcron-controller.go:721]: uncached client here seems wrong as well.

File: pkg/controller/dataimportcron-controller.go (hunk 1 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:708]: Removed. So far no surprises.

File: pkg/controller/dataimportcron-controller.go (hunk 2 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go]: Use your tip, thanks @akalenyu !

File: pkg/controller/dataimportcron-controller.go (hunk 3 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go]: fixed

**Pull request merged**